### PR TITLE
myrescue: init at 0.9.8

### PR DESCRIPTION
**Pull request by D3vil0p3r**
## Description of changes
myrescue: hard disk recovery tool that reads undamaged regions first.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by afh** [comment]
Result of `nixpkgs-review pr 289546` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>myrescue</li>
  </ul>
</details>

**Comment by D3vil0p3r** [comment]
> Result of `nixpkgs-review pr 289546` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 
> 1 package failed to build:

If you have a Darwin machine, can you try to get the source code and build it manually by make? And check if you get an error? So, if we can fix it, we can apply a patch.

**Comment by afh** [comment]
The myrescue package fails to build on darwin due to: `myrescue.c:47:10: fatal error: 'linux/usbdevice_fs.h' file not found`. This seems to suggest that this package might be linux specific? Is that so, @D3vil0p3r?

**Comment by D3vil0p3r** [comment]
> linux/usbdevice_fs.h

The best thing could be to find a similar linux/usbdevice_fs.h header file but for darwin. @a-n-n-a-l-e-e do you have any idea?

**Comment by afh** [comment]
Other things that seem "problematic" and I'm uncertain on how to address are:

* Use of `USBDEVFS_RESET` request with ioctl(2)
* Use of `open64` and `lseek64`

**Comment by a-n-n-a-l-e-e** [comment]
here is a patch that gets darwin to build. didn't fix any of the warnings.
[0001-darwin-build-fixes.patch.gz](https://github.com/NixOS/nixpkgs/files/14321809/0001-darwin-build-fixes.patch.gz)

**Comment by a-n-n-a-l-e-e** [comment]
> Other things that seem "problematic" and I'm uncertain on how to address are:
> 
>     * Use of `USBDEVFS_RESET` request with ioctl(2)
> 
>     * Use of `open64` and `lseek64`

can just use open and lseek on darwin. and defined USBDEVFS_RESET to 0 so not used on darwin.

**Comment by afh** [comment]
Thanks for providing a patch, @a-n-n-a-l-e-e. I can confirm that with the patch applied the package builds and I can run `myrescue -h` successfully.
What are other ways to test the program in order to determine that it functions as expected on macOS?

**Comment by D3vil0p3r** [comment]
@afh I pushed the change. Please can you verify?
@a-n-n-a-l-e-e how do you know what is the right patch file? Is it because you know the darwin libraries and you create a manual patch file?

File: pkgs/by-name/my/myrescue/package.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  sourceRoot = "./src";
+
+  patches = [
+    ./0001-darwin-build-fixes.patch
+  ];
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp myrescue $out/bin/
```
Comment (afh) [inline pkgs/by-name/my/myrescue/package.nix:27]: I've seen several packages use install(1) and it seems this package would benefit from its use too:

```suggestion
    install -Dm755 myrescue -t $out/bin
```

**Comment by afh** [comment]
Result of `nixpkgs-review pr 289546` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>myrescue</li>
  </ul>
</details>

**Comment by D3vil0p3r** [comment]
> Result of `nixpkgs-review pr 289546` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built:

If everything is well, please, approve the PR.

**Review by afh — COMMENTED** [review]
Looks good in general @D3vil0p3r; what are your thoughts on the suggestion below?

File: pkgs/by-name/my/myrescue/package.nix (hunk 1 above)
Comment (D3vil0p3r) [reply pkgs/by-name/my/myrescue/package.nix:27]: @a-n-n-a-l-e-e in general, should we use `mkdir/cp` or `install`?

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/my/myrescue/package.nix:27]: i use install when it makes sense. not sure if there is a policy one way or the other.

**Review by afh — APPROVED** [review]
Well done, @D3vil0p3r, and thanks once more, @a-n-n-a-l-e-e, for helping out with the patch.

I'll leave accepting or disregarding the suggestion about using `install` over `mkdir/cp` at your discretion, @D3vil0p3r 🙂

**Comment by D3vil0p3r** [comment]
> Well done, @D3vil0p3r, and thanks once more, @a-n-n-a-l-e-e, for helping out with the patch.
> 
> I'll leave accepting or disregarding the suggestion about using `install` over `mkdir/cp` at your discretion, @D3vil0p3r 🙂

Thank you for your assistance. I pushed the proposed change.

**Review by ghost — APPROVED** [review]

**Comment by afh** [comment]
If it isn't too much too ask, @D3vil0p3r, and @a-n-n-a-l-e-e agrees I'd appreciate if the following was included in the commit message 🙂 

```
Co-authored-by: Alexis Hildebrandt <afh@surryhill.net>
Co-authored-by: annalee <150648636+a-n-n-a-l-e-e@users.noreply.github.com>
```

File: pkgs/by-name/my/myrescue/0001-darwin-build-fixes.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Subject: [PATCH] darwin build fixes
+
+---
+ compat.h              | 9 +++++++++
+ myrescue-bitmap2ppm.c | 1 +
+ myrescue-stat.c       | 1 +
+ myrescue.c            | 5 +++++
+ 4 files changed, 16 insertions(+)
+ create mode 100644 src/compat.h
+
```
[the remark this answers is not in the document]
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/my/myrescue/0001-darwin-build-fixes.patch:13]: those lines are generated by git when you create a patch. look at many patches in nixpkgs.

**Comment by a-n-n-a-l-e-e** [comment]
> If it isn't too much too ask, @D3vil0p3r, and @a-n-n-a-l-e-e agrees I'd appreciate if the following was included in the commit message 🙂
> 
> ```
> Co-authored-by: Alexis Hildebrandt <afh@surryhill.net>
> Co-authored-by: annalee <150648636+a-n-n-a-l-e-e@users.noreply.github.com>
> ```

i don't think that doing code reviews qualifies as being a co-author.

**Comment by afh** [comment]
How about providing patches, like you did, @a-n-n-a-l-e-e, and offering accepted suggestions, like I did?

If it doesn't feel justified or right, I'm onboard with my Co-authoered request to be discarded 🙂

**Comment by D3vil0p3r** [comment]
> How about providing patches, like you did, @a-n-n-a-l-e-e, and offering accepted suggestions, like I did?
> 
> If it doesn't feel justified or right, I'm onboard with my Co-authoered request to be discarded 🙂

Several Nix mantainers said me to remove all commit messages coming from additional commits during the squash process. Why don't add yourself as maintainer of this package at this point?

**Comment by afh** [comment]
Ah, I wasn't aware of that preference regarding commit messages, thank you for pointing that out, @D3vil0p3r 👍
While I've thoroughly enjoyed helping out on this PR I think I would not be able to do a good job as a maintainer of this package as I do not plan to use it myself (or at least not yet :).

File: pkgs/by-name/my/myrescue/0001-darwin-build-fixes.patch (hunk 2 above)
[the remark this answers is not in the document]
Comment (afh) [reply pkgs/by-name/my/myrescue/0001-darwin-build-fixes.patch:13]: Thanks for the explanation and added context, @a-n-n-a-l-e-e, much appreciated. Yes, git includes those lines when formatting patches, yet the `patch` utility does not need them in order to patch the sources, although I think `git apply` does.

What I've observed in nixpkgs is that superfluous or (overly) verbose information should be stripped before findings its way into nixpkgs, hence my above comment. I could be wrong or the convenience to generate and maintaining patches for nix packages has more value in this case.

If you have links to resources for nixpkgs' best practices do post them, happy to learn 🙂

**Pull request merged**